WIP: Sign systemd boot EFI images for secure booting.

grahamc opened the pull request:
with efitools (`nix-shell -p efitools`)

and running with openssl in the shell environment:

    $ cd /home/grahamc/projects/grahamc/secure-boot
    $ uuidgen --random > GUID.txt
    $ openssl req -newkey rsa:4096 -nodes -keyout PK.key -new -x509 -sha256 -days 3650 -subj "/CN=my Platform Key/" -out PK.crt
    $ openssl x509 -outform DER -in PK.crt -out PK.cer
    $ cert-to-efi-sig-list -g "$(< GUID.txt)" PK.crt PK.esl
    $ sign-efi-sig-list -g "$(< GUID.txt)" -k PK.key -c PK.crt PK PK.esl PK.auth
    $ sign-efi-sig-list -g "$(< GUID.txt)" -c PK.crt -k PK.key PK /dev/null rm_PK.auth

    $ openssl req -newkey rsa:4096 -nodes -keyout KEK.key -new -x509 -sha256 -days 3650 -subj "/CN=my Key Exchange Key/" -out KEK.crt
    $ openssl x509 -outform DER -in KEK.crt -out KEK.cer
    $ cert-to-efi-sig-list -g "$(< GUID.txt)" KEK.crt KEK.esl
    $ sign-efi-sig-list -g "$(< GUID.txt)" -k PK.key -c PK.crt KEK KEK.esl KEK.auth

    $ openssl req -newkey rsa:4096 -nodes -keyout db.key -new -x509 -sha256 -days 3650 -subj "/CN=my Signature Database key/" -out db.crt
    $ openssl x509 -outform DER -in db.crt -out db.cer
    $ cert-to-efi-sig-list -g "$(< GUID.txt)" db.crt db.esl
    $ sign-efi-sig-list -g "$(< GUID.txt)" -k KEK.key -c KEK.crt db db.esl db.auth

and then loading the .auth files in my UEFI firmware, then setting in
my configuration.nix:

    boot = {
      loader = {
        systemd-boot = {
          enable = true;
          signed = true;
          signing-key = "/home/grahamc/projects/grahamc/secure-boot/db.key";
          signing-certificate = "/home/grahamc/projects/grahamc/secure-boot/db.crt";
        };
      };
    };

and running `nixos-rebuild boot` I was able to secure-boot my laptop.

Note: this code signs EVERY boot entry for NixOS, and the BOOT files
from systemd (even if they're already tampered with! since this code
doesn't replace it.) AND does not allow a graceful key phase out. If
the key is replaced in gen2, gen1's boot EFI is deleted and replaced
with a new version.

Also note this only works on x86-64 due to the assumptions in the
builder.py.

**Finally, you can have much fewer generations: each generation on my
laptop takes 20MB in /boot.**

Info taken from:

 - https://wiki.archlinux.org/index.php/Secure_Boot#Using_your_own_keys
 - https://systemd.io/BOOT_LOADER_SPECIFICATION#type-2-efi-unified-kernel-images
 - https://www.freedesktop.org/software/systemd/man/os-release.html

grahamc committed:
Sign systemd boot EFI images for secure booting.

with efitools (`nix-shell -p efitools`)

and running with openssl in the shell environment:

    $ cd /home/grahamc/projects/grahamc/secure-boot
    $ uuidgen --random > GUID.txt
    $ openssl req -newkey rsa:4096 -nodes -keyout PK.key -new -x509 -sha256 -days 3650 -subj "/CN=my Platform Key/" -out PK.crt
    $ openssl x509 -outform DER -in PK.crt -out PK.cer
    $ cert-to-efi-sig-list -g "$(< GUID.txt)" PK.crt PK.esl
    $ sign-efi-sig-list -g "$(< GUID.txt)" -k PK.key -c PK.crt PK PK.esl PK.auth
    $ sign-efi-sig-list -g "$(< GUID.txt)" -c PK.crt -k PK.key PK /dev/null rm_PK.auth

    $ openssl req -newkey rsa:4096 -nodes -keyout KEK.key -new -x509 -sha256 -days 3650 -subj "/CN=my Key Exchange Key/" -out KEK.crt
    $ openssl x509 -outform DER -in KEK.crt -out KEK.cer
    $ cert-to-efi-sig-list -g "$(< GUID.txt)" KEK.crt KEK.esl
    $ sign-efi-sig-list -g "$(< GUID.txt)" -k PK.key -c PK.crt KEK KEK.esl KEK.auth

    $ openssl req -newkey rsa:4096 -nodes -keyout db.key -new -x509 -sha256 -days 3650 -subj "/CN=my Signature Database key/" -out db.crt
    $ openssl x509 -outform DER -in db.crt -out db.cer
    $ cert-to-efi-sig-list -g "$(< GUID.txt)" db.crt db.esl
    $ sign-efi-sig-list -g "$(< GUID.txt)" -k KEK.key -c KEK.crt db db.esl db.auth

and then loading the .auth files in my UEFI firmware, then setting in
my configuration.nix:

    boot = {
      loader = {
        systemd-boot = {
          enable = true;
          signed = true;
          signing-key = "/home/grahamc/projects/grahamc/secure-boot/db.key";
          signing-certificate = "/home/grahamc/projects/grahamc/secure-boot/db.crt";
        };
      };
    };

and running `nixos-rebuild boot` I was able to secure-boot my laptop.

Note: this code signs EVERY boot entry for NixOS, and the BOOT files
from systemd (even if they're already tampered with! since this code
doesn't replace it.) AND does not allow a graceful key phase out. If
the key is replaced in gen2, gen1's boot EFI is deleted and replaced
with a new version.

Also note this only works on x86-64 due to the assumptions in the
builder.py.

Finally, you can have much fewer generations: each generation on my
laptop takes 20MB in /boot.

Info taken from:

 - https://wiki.archlinux.org/index.php/Secure_Boot#Using_your_own_keys
 - https://systemd.io/BOOT_LOADER_SPECIFICATION#type-2-efi-unified-kernel-images
 - https://www.freedesktop.org/software/systemd/man/os-release.html

grahamc commented:
This whole PR was made sort of on a lark, I didn't exactly plan to set out and do this ... but here we are =)

grahamc commented:
@GrahamcOfBorg test boot.uefiUsb

ivegotasthma commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -190,12 +263,19 @@ def main():
     mkdir_p("@efiSysMountPoint@/efi/nixos")
     mkdir_p("@efiSysMountPoint@/loader/entries")
 
+    if @signed@ == 1:
```
I've never seen this python syntax, what is this?

grahamc replied:
This is probably the ugliest part of the whole PR. I'll improve it to make it more clear.

grahamc replied:
Quoted.

ivegotasthma commented:
Very cool concept.

grahamc commented:
That line was actually correct, but resulted in the output in $out/usr/... I've now fixed that, and PR'd efitools separately: https://github.com/NixOS/nixpkgs/pull/54471

ivegotasthma commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -190,12 +263,19 @@ def main():
     remove_old_entries(gens)
     for gen in gens:
-        write_entry(*gen, machine_id)
+        if @signed@ == 1:
```
I think you missed to quote this too.

grahamc replied:
Sure did :) fixed.

Mic92 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +112,74 @@ def write_entry(profile, generation, machine_id):
+    shutil.copyfile(
+        profile_path(profile, generation, "initrd"),
+        initrd
+    )
```
Is this copy really required when it is copied into `linuxx64.efi.stub`?

Mic92 replied:
Ah, ok because of `append-secrets`. Still it would be a good to keep the initrd out of `/boot` to decrease the likely hood of going out of disk-space and because fat32 is not the best filesystem ...

Mic92 replied:
with tempfile.TemporaryDirectory() as tmp:
    # here the code goes

Mic92 replied:
There could be also a check if `append-secrets` or otherwise we don't need to copy at all.

grahamc replied:
Implemented, check it out!

Mic92 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +112,74 @@ def write_entry(profile, generation, machine_id):
+    os.rename(entry_tmp, entry_file)
+    os.unlink(tmp_path)
+    os.unlink(kernel_param_file)
+    os.unlink(initrd)
```
Those should be deleted also in the error case. If you use `TemporaryDirectory` as suggested above, some will be already covered.

grahamc replied:
Indeed, but I haven't made other changes. What is the best way to handle the failure case?

Mic92 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +112,74 @@ def write_entry(profile, generation, machine_id):
+    try:
+        append_initrd_secrets = profile_path(profile, generation, "append-initrd-secrets")
+        subprocess.check_call([append_initrd_secrets, "@efiSysMountPoint@%s" % (initrd)])
+    except FileNotFoundError:
```
I would prefer an explicity check if the file exists:

```
if os.path.exists(append_initrd_secrets):
    subprocess.check_call([append_initrd_secrets, "@efiSysMountPoint@%s" % (initrd)])
```

grahamc replied:
Done!

Mic92 commented:
If you merge this feature could you drop your notes in the manual or the wiki?

grahamc committed:
systemd-boot-builder: don't needlessly copy the initrd

Only copy if we need to append initrd secrets, and even then
don't copy to the EFI partition.

grahamc commented:
Note the whitespace-free diff is a bit easier to see the smaller changes: https://github.com/NixOS/nixpkgs/pull/53901/commits/68f71394a256e480b9ab2e0560012c1e91f7335b?w=1

Yes definitely: docs is a blocker for this PR. But, so is having a few people try it out for a while.

grahamc commented:
PS extending this Python code has not been a lot of fun :(

Mic92 commented:
It gets more fun, once it has type annotations and one has type checks running in its editor because the feedback cycle gets shorter.

Mic92 reviewed (changes requested):
As far as I can see old efi stubs are also not yet removed in `remove_old_entries`.

Mic92 commented:
This `append-initrd` script also looks a bit fishy to me because it appends the same secrets over and over again to existing initrds.

redfish64 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +112,74 @@ def write_entry(profile, generation, machine_id):
+        "@sbsigntool@/bin/sbsign",
+        "--key", "@signingKey@",
+        "--cert", "@signingCertificate@",
+        "--output", output,
```
I rolled my own version of this using the same sources and ran into a problem that if the disk fills up while sbsign is writing it's output, it will return 0 (success) regardless. This is a big problem, because:
 
a) Since each efi now contains its own separate initrd and by default old versions aren't cleaned up, the boot disk can fill up quite quickly, and
b) The code signs *all* versions. So as soon as the disk does fill up, all generations fail to be signed and are mangled silently by sbsign.

This causes the boot process to have appeared to work, but result in system where several, possibly all generations are unbootable. 

I'd recommend adding the following after the sbsign command to verify the binary was signed correctly before moving on:

    subprocess.check_call([
        "@sbsigntool@/bin/sbverify",
        "--cert", "@signingCertificate@",
        output
    ])

This way the process will error out, and the user will at least get some sort of indication that something went wrong.

grahamc replied:
Good grief, @redfish64! Thank you!

grahamc replied:
I have extended sign_path to sign to a tmpdir, move, verify, move.

grahamc commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +112,74 @@ def write_entry(profile, generation, machine_id):
+                fp.write("machine-id %s\n" % machine_id)
+
+        os.rename(entry_tmp, entry_file)
+        os.unlink(tmp_path)
```
rename doesn't work because it is a different filesystem.

Mic92 committed:
Update nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

Co-Authored-By: grahamc <graham@grahamc.com>

Mic92 committed:
Update nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

Co-Authored-By: grahamc <graham@grahamc.com>

Mic92 committed:
Update nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

Co-Authored-By: grahamc <graham@grahamc.com>

Mic92 committed:
Update nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

Co-Authored-By: grahamc <graham@grahamc.com>

Mic92 committed:
Update nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

Co-Authored-By: grahamc <graham@grahamc.com>

Mic92 committed:
Update nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

Co-Authored-By: grahamc <graham@grahamc.com>

grahamc commented:
> As far as I can see old efi stubs are also not yet removed in remove_old_entries.

NixOS removes all of its own files on each switch, so this is already solved.

> This append-initrd script also looks a bit fishy to me because it appends the same secrets over and over again to existing initrds. Also this only affect the non-secure boot case at the moment.

also a non-issue, since those files are "new" on each run.

grahamc commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -190,12 +263,19 @@ def main():
     for gen in gens:
-        write_entry(*gen, machine_id)
+        if "@signed@"== "1":
+            write_secureboot_entry(*gen, machine_id)
```
See my reply: https://github.com/NixOS/nixpkgs/pull/53901#issuecomment-456803888

grahamc committed:
system-boot-builder: verify signed efi programs before moving them in to place

grahamc commented:
I was wrong -- it doesn't delete every file... I'm not sure where I got that idea. Anyway, I fixed it so the .efi's aren't deleted on each run.

grahamc committed:
system-boot-builder: pyflakes fixups

grahamc committed:
system-boot-builder: don't re-sign/re-build initrds

grahamc commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +110,105 @@ def write_entry(profile, generation, machine_id):
+    efi_file = "@efiSysMountPoint@/%s" % (efi_file_relative)
+
+    try:
+        sbverify(efi_file)
```
This is not a valid/safe statement, because:

```
$ sudo rm /nix/var/nix/profiles/system-37-link # latest
$ sudo nixos-rebuild boot
$ sudo rm /nix/var/nix/profiles/system-37-link # latest
$ sudo nixos-rebuild boot --upgrade
$ nix-collect-garbage
```

will make boot option 37 unbootable. Maybe storing it by the hash part of the generation's store path root would work better.

grahamc commented:
@GrahamcOfBorg eval

Mic92 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +110,105 @@ def write_entry(profile, generation, machine_id):
+
+    try:
+        sbverify(efi_file)
+    except:
```
Better to list the exception explicitly i.e. `OSError`. A bare except will also catch syntax errors.

Mic92 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -107,6 +110,105 @@ def write_entry(profile, generation, machine_id):
+        try:
+            sbverify(f"{output}.tmp")
+            os.rename(f"{output}.tmp", output)
+        except:
```
```suggestion
        except subprocess.CalledProcessError:
```

florianjacob commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -41,6 +47,38 @@ in {
+      description = ''
+        The <literal>db.key</literal> signing key, for signing EFI
+        programs. Note: Do not pass a store path. Passing the key like
+        <literal>signing-key = ./db.key;</literal> will copy the
```
Couldn't this hazard be avoided by using `toString`, i.e. `signingKey = if cfg.signed then (toString cfg.signing-key) else "/no-signing-key";`? Learned this from [this stackoverflow answer](https://stackoverflow.com/questions/43850371/when-does-a-nix-path-type-make-it-into-the-nix-store-and-when-not)

monokrome commented:
:exclamation: SO EXCITE :exclamation:

bbigras commented:
Any progress on this?

Mic92 commented:
I think Graham wanted some people to test this code before he would merge it.

tkerber commented:
I'm now testing this on my machines, and so far it's working perfectly.

One idea I did have to maybe make the process more straightforward is to by default have a location for the certificates and keys in `/var`, which a oneshot service populates if it does not exist (somewhat like how `letsencrypt` certificates are automatically requisitioned). That said, users would still need to know where to find certs, and how to install them in their BIOS, so I'm not sure if it's worth it.

eadwu commented:
Working fine on my 9570.

Mic92 commented:
@tkerber generating certificates sounds like a good idea. The chain of commands looks complex enough to make it a timer saver.

giacomocariello commented:
I've recently tested on HP Microserver Gen10 with positive result. sbsign whines several times that "gaps between PE/COFF sections?", but apparently it's not a problem.

Mic92 commented:
If @grahamc does have no time to work on this PR, somebody else can take it over. I would review it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/cli-to-plasma-installation-docs-or-help/4631/6

makefu commented:
I'd love to see this PR upstream, i wanted to set up my laptop with secure boot for a long time

bbigras commented:
Any progress on this?

grahamc commented:
Taking a look.

lhindir commented:
@grahamc Any results from your look? Thanks!

bbigras commented:
Friendly reminder to look at this again. :smiley:

GrahamcOfBorg commented:
I took another look. I got stumped at wanting to test this automatically in a VM test and I wasn’t sure how to generate keys and register it with Qemu. 

On Sat, Apr 11, 2020, at 2:15 PM, Bruno Bigras wrote:
> 

> Friendly reminder to look at this again. 😃


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NixOS/nixpkgs/pull/53901#issuecomment-612481616>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGCEZ2VQ7EWAZUSOV7LV6RTRMCXVTANCNFSM4GPVW2YA>.

danielfullmer commented:
@grahamc I submitted a PR for testing `systemd-boot`, which required fixing some issues with installing the bootloader in the testing environment. It might prove useful for testing secure boot as well.

grahamc commented:
Just reposting as me (oops):

I took another look. I got stumped at wanting to test this automatically in a VM test and I wasn’t sure how to generate keys and register it with Qemu.

danielfullmer commented:
@grahamc I had some free time this weekend and decided to take a stab at making a test for this: https://github.com/danielfullmer/nixpkgs/commit/a2b95a3caa79f05dbab4ef06f7182bf7d4725099

To easily test:
```
nix-build "https://github.com/danielfullmer/nixpkgs/archive/a2b95a3caa79f05dbab4ef06f7182bf7d4725099.tar.gz" -A nixosTests.systemd-boot.secureBoot
```

It uses [ovmf-vars-generator](https://github.com/rhuefi/qemu-ovmf-secureboot) to load a certificate into `nvram` and enable secure boot. It still definitely needs some reorganization, but I'm out of time today and I finally got it working--so I thought I might share. :)

grahamc commented:
This is amazing! Thank you a lot, @danielfullmer. That should make it a lot easier to get a test going! :grin:

bbigras commented:
any progress on this awesome feature?

shadowrylander commented:
I'm assuming the `systemd-boot` signed options aren't available yet?

lheckemann commented:
@shadowrylander they are if you use the PR branch.

shadowrylander commented:
@lheckemann Could you give me a code example? I'm still fairly new to NixOS, and having trouble understanding the syntax.

lheckemann commented:
You could use the current state of the branch, but that comes with the disadvantage that it's over a year old now and you'll get over-a-year-old software (without security patches from that period) correspondingly. If you're willing to put the effort in to rebase it on current master (which will take at least a basic understanding of nix syntax so you can work out how to fix the conflicts), you can then build your system from it by passing `-I nixpkgs=$path/to/your/nixpkgs` to `nixos-rebuild`. Otherwise no, it's not available yet.

shadowrylander commented:
Hmm... Well, it should be a great way to learn the syntax, at the very least; I think I'll try it. Thanks!

bbigras commented:
Friendly ping

bbigras commented:
Anyone has the skills to pick up this PR?

jsimonetti commented:
I can add the code from this PR still works fine on 20.09 when used as a module.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-on-secure-boot-debian-based-laptop/10976/2

hmenke commented:
@danielfullmer Do you feel like taking this forward?  It seems that @grahamc doesn't have time to push this anymore.  I'm actually using this code for almost a year already and it would be really great if it made into 21.05.

06kellyjac commented:
As an observer, what extra work is required to make this ready? (Excluding resolving merge conflicts)

GrahamcOfBorg commented:
I think for something so security critical it needs an automatic VM test. I think that is pretty much it.

On Sat, Jan 16, 2021, at 1:22 PM, Jack Kelly wrote:
> 

> As an observer, what extra work is required to make this ready? (Excluding resolving merge conflicts)


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NixOS/nixpkgs/pull/53901#issuecomment-761610801>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGCEZ2WQ7AHHSYX74VCXMTTS2HKPHANCNFSM4GPVW2YA>.

danielfullmer commented:
@hmenke I unfortunately can't commit to having time to finish up this PR.  I should be available to answer any questions about the tests if anyone else would like to finish this up.  I'd also like to add that we'd also need some documentation for this feature.

However, I do have a slightly more recent version of the secure boot tests at: https://github.com/danielfullmer/nixpkgs/tree/secure-boot-wip2
The biggest change is that the logic for adding secure boot certs was moved into `qemu-vm.nix`.  IIRC, `nixos-rebuild build-vm-with-bootloader` should now also be able to pre-enroll the certificate specified under `boot.loader.systemd-boot.signing-certificate`, so you could interactively test secure boot functionality.

I might not be the best to finish up this PR since I personally wasn't planning to use this feature on my machine until we also had working kernel module signing (https://github.com/NixOS/nixpkgs/pull/87426) and had chained integrity checking to the actual generation NixOS boots via dm-verity / fs-verity / IMA, or something similar, for which I don't have any great ideas.

kisik21 commented:
WOW! I kinda wanted to have this and now it exists.

I don't think the security would suffer too much from the fact that files are re-signed on every `nixos-rebuild boot`, if the files are re-installed from the /nix/store every time. Then, if the system booted successfully with Secure Boot from the beginning, and the /nix/store wasn't tampered with, I don't think there's a chance someone untrusted could swap the bootloader. Quite the contrary, every rebuild will ensure the system is in bootable state and all the files are trusted.

cwyc commented:
Thanks a lot for this module.
I've managed to merge it with nixos 21.05:
https://github.com/cwyc/nixpkgs/tree/secure-boot-rebased-21.05/nixos/modules/system/boot/loader/systemd-boot
And I've adapted it as an importable module, with the options renamed to `boot.loader.systemd-boot-secure`:
https://gist.github.com/cwyc/9de83566c92bdabd000256ed49312a53

bbigras commented:
@cwyc Are you going to open a PR? :pray:

cwyc commented:
I've also tried out @danielfullmer 's [tests](https://github.com/cwyc/nixpkgs/commit/e47272ddc958800311eb6f9be975b1e2b879e610). They work on my end, what is it that has to be changed?

bbigras commented:
Maybe open a new PR and ask for reviews again.

mweinelt commented:
Please, please, if someone has the time and knowledge to take this over the finish line, please do. We will apparently not get very many testers without this having landed in nixpkgs.

benpye commented:
@mweinelt I'm definitely interested in seeing this land - what's still missing from the changes that @cwyc posted above? Whilst as mentioned earlier without having chained verification of kernel modules etc we don't get full trust, this is definitely a significant stepping stone to get there.

TLATER commented:
@benpye I think a start would be addressing the review comments - there are plenty right there, some really low hanging fruit!

Baughn commented:
> 
> 
> Please, please, if someone has the time and knowledge to take this over the finish line, please do. We will apparently not get very many testers without this having landed in nixpkgs, and the value of having this built-in and easily set up is so huge.

It's something that really ought to be merged and functional by the time of the next release, because a lot of people are about to lose the option of running their systems with secure boot disabled.

Given Windows 11 has it as a requirement, any operating system which doesn't support it at all can't boot. Not without reconfiguring your BIOS at every reboot, at any rate, which I don't think many people are going to do. Personally I've been running an indev version of 11, and...

Well, it's a good upgrade. With the exception of the usual development snapshot bugs, most everything works better than on 10.

I don't expect anyone who's already using NixOS to be stopped by this, so it might not seem that important, but losing the ability to dual-boot will make it far harder to make new converts.

grahamc commented:
> lose the option of running their systems with secure boot disabled.

> not without reconfiguring your BIOS at every reboot

This sounds like FUD to me. I don't see why Windows11 making it a requirement makes the hardware or firmware behave differently than it has for years.

I have a different approach to this PR in progress, but I don't think this particular implementation is good or worth merging: the lifecycle around signing is unfit for purpose. To that end, I'm closing this because I would be -1 on an implementation similar to this merging.

cwyc commented:
@Baughn 
Looks like it's only an issue if you want to dual-boot with Windows 11. But if you are dual-booting, surely you can't replace Microsoft's keys with the self-signing keys used here, a secure boot solution would have to use shim/preloader.

matshch commented:
> @Baughn
> surely you can't replace Microsoft's keys with the self-signing keys used here

You can import several signing keys in UEFI. This way, I have my own Platform Key, two Key Exchange Keys (my own and Microsoft's, so Windows can autoupdate dbx), two keys in db (again, my own and Microsoft's), and dbx is full of Microsoft's revoked signs. This way, I have Secure dual-boot with Windows 10 and Ubuntu (autosigned by https://github.com/julian-klode/sicherboot).

Installing NixOS requires disabling Secure Boot, which will somewhat lower my security, and this is one of the few reasons that stops me from daily driving NixOS.

ZerataX commented:
> Given Windows 11 has it as a requirement

that only seems to be true for installation. I installed win 11 then turned off secure boot and am now back to dual booting win 11 and nixos

TLATER commented:
@grahamc do please link your solution here once it's in a state worth looking at so we get a notification :)

eadwu commented:
I don't know if Windows 11 has it as a requirement but I have been using it since release to Dev and the like past 3 updates and have not gotten anything needing Secure Boot to be re-enabled (whether during upgrade process or daily driving).

Baughn commented:
The source would be https://blogs.windows.com/windows-insider/2021/06/28/update-on-windows-11-minimum-system-requirements/

It's true that at the moment you can disable secure boot after installation, and it'll keep running, but I don't think it's especially safe to assume that will still be the case with the final release. I've never seen any statement from MS saying it will be, and I don't think it'd be safe to assume even if I had.

grahamc commented:
Okay. Good to confirm there is no imminent risk here.

frogamic commented:
I have been meaning to implement an alternative to this myself for a while but I finally got round to it. This completely replaces the configuration file approach with unified kernel images.

https://github.com/frogamic/nix-machines/tree/main/modules/systemd-secure-boot

Some caveats:
* signing is not optional in this script, though `sign_if_not_signed` could easily be replaced with different behavior if signing is not enabled, sticking with unified images by default for `systemd-boot`
* doesn't sign memtest
* doesn't sign fwupd bootloader as used for platform firmware updates
* Leaves behind all the old initrd/bzimage files in `EFI/Nixos`, should the eventual script nuke these every time on the off-chance they exist?

Some thoughts I had while doing it:
* Could the unified image be built as a nix derivation instead? Probably in the `extraSystemBuildCmds` since not all boot methods need it?
  * What happens in that case if a generation doesn't have the unified image? is it just skipped?
* Should secureboot be a configuration option on the same level as `systemd-boot` rather than inside it so all boot methods could share it as an option, and any other efi binaries (eg `fwupdx64.efi`) would use it too?
  * can you even sign your own grub efi like you can with systemd-boot? I never looked into it at all

pnmadelaine commented:
I've used @frogamic approach on my laptop and I'd really like to see this in NixOS.
Is anyone working on a new PR at the moment?

A few comments I have:
- I'd say defaulting to unified images is not a great idea, since it takes a lot more space
- maybe splitting this in two options would be better, the secure boot option requiring to enable the unified images option?
- a nice to have would be an option to build efitools' LockDown.efi with the user's certificates and add it to the boot menu

grahamc commented:
>  Is anyone working on a new PR at the moment?

Yes, my company determinate.systems has been investing considerable time and money in to a suitable implementation over the past two months. We're about ready to publish some WIP code. Typically we do all our work in public, however I wanted to have lots of room to experiment and try different approaches. We're also in the process of applying for funding to implement a few workflows.
